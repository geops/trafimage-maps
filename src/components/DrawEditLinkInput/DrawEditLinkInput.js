import React, { useState, useMemo, useEffect } from 'react';
import { makeStyles, Typography } from '@material-ui/core';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import qs from 'query-string';
import { ReactComponent as Loader } from '../../img/loader.svg';
import PermalinkInput from '../PermalinkInput';
import { DRAW_PARAM, MAPSET_PARENT_PARAM } from '../../utils/constants';

const useStyles = makeStyles(() => ({
  root: {
    position: 'relative',
  },
  loader: {
    position: 'absolute',
    top: 0,
    height: '100%',
    display: 'flex',
    alignItems: 'center',
  },
}));

function DrawEditLinkInput() {
  const classes = useStyles();
  const { t } = useTranslation();
  const [isLoading, setLoading] = useState(true);
  const [shortenUrl, setShortenUrl] = useState('');
  const mapsetUrl = useSelector((state) => state.app.mapsetUrl);
  const drawIds = useSelector((state) => state.app.drawIds);
  const shortenerUrl = useSelector((state) => state.app.shortenerUrl);

  // Get the mapset url
  const url = useMemo(() => {
    const params = qs.parse(window.location.search);
    params[DRAW_PARAM] = drawIds.admin_id;
    return `${mapsetUrl}?${MAPSET_PARENT_PARAM}=${encodeURIComponent(
      `${window.location.href.split('?')[0]}?${qs.stringify(params)}`,
    )}`;
  }, [drawIds, mapsetUrl]);

  // Get the shortened url
  useEffect(() => {
    if (!shortenerUrl) {
      return () => {};
    }
    setLoading(true);
    const abortController = new AbortController();
    fetch(`${shortenerUrl}?url=${url}`, { signal: abortController.signal })
      .then((res) => res.json())
      .then((data) => {
        if (data.error) {
          throw new Error(data.error);
        }
        setLoading(false);
        setShortenUrl(data.url);
      })
      .catch(() => {
        setLoading(false);
        setShortenUrl(url);
      });
    return () => {
      abortController.abort();
    };
  }, [url, shortenerUrl]);

  if (!drawIds || !drawIds.admin_id) {
    return null;
  }

  return (
    <div className={classes.root}>
      <PermalinkInput value={shortenUrl} />
      {isLoading && (
        <Typography variant="subtitle1">
          <span className={classes.loader}>
            <Loader /> <span>{t('Laden...')}</span>
          </span>
        </Typography>
      )}
    </div>
  );
}

export default React.memo(DrawEditLinkInput);
