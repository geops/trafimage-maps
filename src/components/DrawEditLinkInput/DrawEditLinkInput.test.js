import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import DrawEditLinkInput from '.';

describe('DrawEditLinkInput', () => {
  const mockStore = configureStore([thunk]);
  let store;
  test('should match snapshot.', () => {
    store = mockStore({
      map: {},
      app: { drawIds: {} },
    });
    const component = renderer.create(
      <Provider store={store}>
        <DrawEditLinkInput />
      </Provider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
