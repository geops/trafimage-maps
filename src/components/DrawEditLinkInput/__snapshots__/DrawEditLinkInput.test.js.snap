// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawEditLinkInput should match snapshot. 1`] = `null`;
